applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py

```diff
@@ -0,0 +1,760 @@
+        self.main_model_part.ProcessInfo[KratosMultiphysics.STEP] += 1
+        self.main_model_part.CloneTimeStep(new_time)
+
+        return new_time
+
+    def ComputeDeltaTime(self):
+        return self.settings["time_stepping"]["time_step"].GetDouble()
+
+    def GetComputingModelPart(self):
+        return self.main_model_part.GetSubModelPart(self.settings["computing_model_part_name"].GetString())
+
+    def ExportModelPart(self):
+        name_out_file = self.settings["model_import_settings"]["input_filename"].GetString()+".out"
+        file = open(name_out_file + ".mdpa","w")
+        file.close()
+        KratosMultiphysics.ModelPartIO(name_out_file, KratosMultiphysics.IO.WRITE).WriteModelPart(self.main_model_part)
+
+    def SetEchoLevel(self, level):
+        self.get_convection_diffusion_solution_strategy().SetEchoLevel(level)
+
+    def Clear(self):
+        self.get_convection_diffusion_solution_strategy().Clear()
+
+    def Check(self):
+        self.get_convection_diffusion_solution_strategy().Check()
+
+    #### Specific internal functions ####
+
+    def get_solution_scheme(self):
+        if not hasattr(self, '_solution_scheme'):
+            self._solution_scheme = self._create_solution_scheme()
+        return self._solution_scheme
+
+    def get_convergence_criterion(self):
+        if not hasattr(self, '_convergence_criterion'):
+            self._convergence_criterion = self._create_convergence_criterion()
+        return self._convergence_criterion
+
+    def get_linear_solver(self):
+        if not hasattr(self, '_linear_solver'):
+            self._linear_solver = self._create_linear_solver()
+        return self._linear_solver
+
+    def get_builder_and_solver(self):
+        if not hasattr(self, '_builder_and_solver'):
+            self._builder_and_solver = self._create_builder_and_solver()
+        return self._builder_and_solver
+
+    def get_convection_diffusion_solution_strategy(self):
+        if not hasattr(self, '_convection_diffusion_solution_strategy'):
+            self._convection_diffusion_solution_strategy = self._create_convection_diffusion_solution_strategy()
+        return self._convection_diffusion_solution_strategy
+
+    def get_restart_utility(self):
+        if not hasattr(self, '_restart_utility'):
+            self._restart_utility = self._create_restart_utility()
+        return self._restart_utility
+
+    def import_materials(self):
+        materials_filename = self.settings["material_import_settings"]["materials_filename"].GetString()
+        if (materials_filename != ""):
+            import read_materials_process
+            # Create a dictionary of model parts.
+            Model = KratosMultiphysics.Model()
+            Model.AddModelPart(self.main_model_part)
+            # Add constitutive laws and material properties from json file to model parts.
+            read_materials_process.ReadMaterialsProcess(Model, self.settings["material_import_settings"])
```

Comment:
cannot you pass `self.model`?

Reply:
yes, actually we should NEVER (or almost so) create a new model...

Reply:
I will check, maybe is a leftover from the first implementation